File: src/core/wheel.js
/** European roulette wheel layout, sector definitions, and position utilities. */

export const WHEEL_ORDER = [
  0, 32, 15, 19, 4, 21, 2, 25, 17, 34, 6, 27, 13, 36, 11, 30,
  8, 23, 10, 5, 24, 16, 33, 1, 20, 14, 31, 9, 22, 18, 29, 7,
  28, 12, 35, 3, 26
];

export const TOTAL_POCKETS = WHEEL_ORDER.length; // 37

export const NUMBER_TO_POSITION = new Map(WHEEL_ORDER.map((n, i) => [n, i]));
export const POSITION_TO_NUMBER = new Map(WHEEL_ORDER.map((n, i) => [i, n]));
export const NUMBER_TO_ANGLE = new Map(WHEEL_ORDER.map((n, i) => [n, (i / TOTAL_POCKETS) * 360]));

export const RED_NUMBERS = new Set([1,3,5,7,9,12,14,16,18,19,21,23,25,27,30,32,34,36]);
export const BLACK_NUMBERS = new Set([2,4,6,8,10,11,13,15,17,20,22,24,26,28,29,31,33,35]);

export function numberColor(n) {
  if (n === 0) return 'green';
  return RED_NUMBERS.has(n) ? 'red' : 'black';
}

export const SECTORS = {
  A: WHEEL_ORDER.slice(0, 6),
  B: WHEEL_ORDER.slice(6, 12),
  C: WHEEL_ORDER.slice(12, 18),
  D: WHEEL_ORDER.slice(18, 24),
  E: WHEEL_ORDER.slice(24, 30),
  F: WHEEL_ORDER.slice(30, 37),
};

export const NUMBER_TO_SECTOR = new Map();
for (const [name, nums] of Object.entries(SECTORS)) {
  for (const n of nums) NUMBER_TO_SECTOR.set(n, name);
}

export function getSector(number) {
  return NUMBER_TO_SECTOR.get(number);
}

export function getNeighbors(number, count = 2) {
  const pos = NUMBER_TO_POSITION.get(number);
  const neighbors = [];
  for (let offset = -count; offset <= count; offset++) {
    if (offset === 0) continue;
    const neighborPos = ((pos + offset) % TOTAL_POCKETS + TOTAL_POCKETS) % TOTAL_POCKETS;
    neighbors.push(POSITION_TO_NUMBER.get(neighborPos));
  }
  return neighbors;
}

export function wheelDistance(a, b) {
  const posA = NUMBER_TO_POSITION.get(a);
  const posB = NUMBER_TO_POSITION.get(b);
  const diff = Math.abs(posA - posB);
  return Math.min(diff, TOTAL_POCKETS - diff);
}

export const BALL_SIZES = ['small', 'medium', 'large'];
export const SPIN_SPEEDS = ['slow', 'medium', 'fast'];
export const WHEEL_SPEEDS = ['slow', 'medium', 'fast'];
export const BALL_DIRECTIONS = ['cw', 'ccw'];

File: src/wheel.py
"""European roulette wheel layout, sector definitions, and position utilities."""

# Physical wheel order (clockwise from 0) — European single-zero
WHEEL_ORDER = [
    0, 32, 15, 19, 4, 21, 2, 25, 17, 34, 6, 27, 13, 36, 11, 30,
    8, 23, 10, 5, 24, 16, 33, 1, 20, 14, 31, 9, 22, 18, 29, 7,
    28, 12, 35, 3, 26
]

TOTAL_POCKETS = len(WHEEL_ORDER)  # 37

# Number -> wheel position index (0-36)
NUMBER_TO_POSITION = {num: idx for idx, num in enumerate(WHEEL_ORDER)}

# Position index -> number
POSITION_TO_NUMBER = {idx: num for idx, num in enumerate(WHEEL_ORDER)}

# Number -> angle in degrees on the wheel
NUMBER_TO_ANGLE = {num: (idx / TOTAL_POCKETS) * 360 for idx, num in enumerate(WHEEL_ORDER)}

# Number -> color
RED_NUMBERS = {1, 3, 5, 7, 9, 12, 14, 16, 18, 19, 21, 23, 25, 27, 30, 32, 34, 36}
BLACK_NUMBERS = {2, 4, 6, 8, 10, 11, 13, 15, 17, 20, 22, 24, 26, 28, 29, 31, 33, 35}


def number_color(n: int) -> str:
    if n == 0:
        return "green"
    return "red" if n in RED_NUMBERS else "black"


# 6 sectors based on physical wheel position
SECTORS = {
    "A": WHEEL_ORDER[0:6],    # [0, 32, 15, 19, 4, 21]
    "B": WHEEL_ORDER[6:12],   # [2, 25, 17, 34, 6, 27]
    "C": WHEEL_ORDER[12:18],  # [13, 36, 11, 30, 8, 23]
    "D": WHEEL_ORDER[18:24],  # [10, 5, 24, 16, 33, 1]
    "E": WHEEL_ORDER[24:30],  # [20, 14, 31, 9, 22, 18]
    "F": WHEEL_ORDER[30:37],  # [29, 7, 28, 12, 35, 3, 26]
}

# Reverse lookup: number -> sector name
NUMBER_TO_SECTOR = {}
for sector_name, numbers in SECTORS.items():
    for num in numbers:
        NUMBER_TO_SECTOR[num] = sector_name


def get_sector(number: int) -> str:
    """Return the sector name for a given number."""
    return NUMBER_TO_SECTOR[number]


def get_neighbors(number: int, n: int = 2) -> list[int]:
    """Return n neighbors on each side of a number on the physical wheel."""
    pos = NUMBER_TO_POSITION[number]
    neighbors = []
    for offset in range(-n, n + 1):
        if offset == 0:
            continue
        neighbor_pos = (pos + offset) % TOTAL_POCKETS
        neighbors.append(POSITION_TO_NUMBER[neighbor_pos])
    return neighbors


def wheel_distance(a: int, b: int) -> int:
    """Minimum arc distance (in pockets) between two numbers on the wheel."""
    pos_a = NUMBER_TO_POSITION[a]
    pos_b = NUMBER_TO_POSITION[b]
    diff = abs(pos_a - pos_b)
    return min(diff, TOTAL_POCKETS - diff)
